org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacCompiler.java

```diff
@@ -109,13 +116,12 @@ public void compile(ICompilationUnit[] sourceUnits) {
 			.collect(Collectors.groupingBy(this::computeOutputDirectory));
 
 		// Register listener to intercept intermediate results from Javac task.
-		JavacTaskListener javacListener = new JavacTaskListener(this.compilerConfig, outputSourceMapping, this.problemFactory);
+		JavacTaskListener javacListener = new JavacTaskListener(this.compilerConfig, outputSourceMapping, this.problemFactory, this.fileObjectToCUMap);
 		MultiTaskListener mtl = MultiTaskListener.instance(javacContext);
 		mtl.add(javacListener);
 
 		for (Entry<IContainer, List<ICompilationUnit>> outputSourceSet : outputSourceMapping.entrySet()) {
-			// Configure Javac to generate the class files in a mapped temporary location
-			var outputDir = JavacClassFile.getMappedTempOutput(outputSourceSet.getKey()).toFile();
+			var outputDir = javaProject.getProject().getParent().findMember(outputSourceSet.getKey().getFullPath()).getLocation().toFile();
```

Comment:
This will break the incremental build. When Javac writes the class files to the target directory directly, the incremental builder cannot get the delta of class changes.